ui/src/lib/gql/job/useSubscriptions.ts

```diff
@@ -0,0 +1,125 @@
+import { useQuery, useQueryClient } from "@tanstack/react-query";
+import { useRef, useEffect, useCallback } from "react";
+
+import { OnJobStatusChangeSubscription } from "../__gen__/graphql";
+import { toJobStatus } from "../convert";
+import { useWsClient } from "../provider/GraphQLSubscriptionProvider";
+
+import { JobQueryKeys } from "./useQueries";
+
+const JOB_STATUS_SUBSCRIPTION = `
+ subscription OnJobStatusChange($jobId: ID!) {
+   jobStatus(jobId: $jobId)
+ }
+`;
+
+export const useJobStatus = (jobId: string) => {
+  const wsClient = useWsClient();
+  const queryClient = useQueryClient();
+  const isSubscribedRef = useRef(false);
+  const unSubscribedRef = useRef<(() => void) | undefined>(undefined);
+  const lastStatusRef = useRef<any>(null);
+
+  const query = useQuery({
+    queryKey: [JobQueryKeys.GetJobStatus, jobId],
+    queryFn: async () => {
+      const cachedData = queryClient.getQueryData([
+        JobQueryKeys.GetJobStatus,
+        jobId,
+      ]);
+      if (cachedData) {
+        lastStatusRef.current = cachedData;
+        return cachedData;
+      }
+      return null;
+    },
+    // Important: initial query should run only once
+    staleTime: Infinity,
+    gcTime: Infinity,
+    refetchOnWindowFocus: false,
+    refetchOnMount: false,
+    refetchOnReconnect: false,
+  });
+
+  useEffect(() => {
+    if (!jobId || isSubscribedRef.current) return;
+
+    console.log(`Setting up job status subscription for job: ${jobId}`);
+    isSubscribedRef.current = true;
+
+    // Initialize with any cached data
+    const cachedData = queryClient.getQueryData([
+      JobQueryKeys.GetJobStatus,
+      jobId,
+    ]);
+    if (cachedData) {
+      lastStatusRef.current = cachedData;
+    }
+
+    // Subscribe to job status
+    const unsubscribe = wsClient.subscribe<OnJobStatusChangeSubscription>(
+      {
+        query: JOB_STATUS_SUBSCRIPTION,
+        variables: { jobId },
+      },
+      {
+        next: (data) => {
+          if (data.data?.jobStatus) {
+            const newStatus = data.data.jobStatus;
+
+            const currentStatus = JSON.stringify(lastStatusRef.current);
+            const incomingStatus = JSON.stringify(newStatus);
+
+            if (currentStatus !== incomingStatus) {
+              console.log(`Job status changed for ${jobId}:`, {
+                previous: lastStatusRef.current,
+                new: newStatus,
+              });
+
+              lastStatusRef.current = newStatus;
+
+              // Update React Query cache
+              queryClient.setQueryData(
+                [JobQueryKeys.GetJobStatus, jobId],
+                toJobStatus(newStatus),
+              );
+            } else {
+              console.log(
+                `Received same status ${newStatus}, no update needed`,
+              );
+            }
+          }
+        },
+        error: (err) => {
+          console.error(`Status subscription error ${jobId}:`, err);
+          isSubscribedRef.current = false;
```

Comment:
I'm not so sure an error breaks the subscription, so this might produce unexpected behaviour